src/btree/bt_cursor.c

```diff
@@ -55,15 +55,31 @@ __cursor_state_restore(WT_CURSOR *cursor, WT_CURFILE_STATE *state)
  *	Return if we have a page pinned.
  */
 static inline bool
-__cursor_page_pinned(WT_CURSOR_BTREE *cbt, bool eviction_ok)
+__cursor_page_pinned(WT_CURSOR_BTREE *cbt)
 {
+	WT_SESSION_IMPL *session;
+
+	session = (WT_SESSION_IMPL *)cbt->iface.session;
+
+	if (!F_ISSET(cbt, WT_CBT_ACTIVE))
+		return (false);
+
 	/*
 	 * Optionally fail the page-pinned test when the page is flagged for
 	 * forced eviction (so we periodically release pages grown too large).
 	 * The test is optional as not all callers can release pinned pages.
+	 *
+	 * We cannot use a page without history for an update operation:
+	 * release the page so we get it again with history in that case.
 	 */
-	return (F_ISSET(cbt, WT_CBT_ACTIVE) &&
-	    (!eviction_ok || cbt->ref->page->read_gen != WT_READGEN_OLDEST));
+	if (cbt->ref->page->read_gen == WT_READGEN_OLDEST)
+		return (false);
+
+	if (cbt->ref->state == WT_REF_LIMBO &&
+	    F_ISSET(&session->txn, WT_TXN_UPDATE))
+		return (false);
```

Comment:
@sueloverso, if this is changed to:

```
	if (F_ISSET(&session->txn, WT_TXN_UPDATE) &&
            cbt->ref->state != WT_REF_MEM)
		return (false);
```

Does that address the issue with attempting to update a LIMBO page due to a transient change in the ref state?